In [1]:
import qiskit
from qiskit_ibm_runtime import QiskitRuntimeService
from qiskit import QuantumRegister, ClassicalRegister, QuantumCircuit

## Connect to IBM Quantum By; Creating IBMid

In [2]:
service = QiskitRuntimeService(channel="ibm_cloud", instance="crn:v1:bluemix:public:quantum-computing:us-east:a/3fd1683bdb2f4bd7b5a9c16264e9531f:ad7d91c7-83ac-4940-9232-867c319705d5::", token="A739W1bW-U5UuPhvQom6N87S-N5DA3tf2o8YpfAeS_bR")

## Part 1
### Create An Empty Circuit
- Initialize the number of qubits
- Intialize the number of cbits

In [3]:
qubits = QuantumRegister(2)
cbits = ClassicalRegister(2)

In [11]:
circuit = QuantumCircuit(qubits, cbits)

In [12]:
circuit.draw()

q0_0: 
      
q0_1: 
      
c0: 2/

## Part 2
- To place qubit 0 in the |+> state and qubit 1 in the |0> state:


In [13]:
#plus state
circuit.h(qubits[0])

In [14]:
#|0> state
circuit.id(qubits[1])

## Part 3
#### A CNOT gate to the circuit from part 2 where qubit 0 is the control bit and qubit 1 is the target bit. 

In [15]:
circuit.cx(qubits[0], qubits[1])

## Part 4

### Add two measurement gates

In [16]:
# Add measurement gates to the circuit
circuit.measure(qubits[0], cbits[0])
circuit.measure(qubits[1], cbits[1])


In [17]:
circuit.draw()

┌───┐     ┌─┐   
q0_0: ┤ H ├──■──┤M├───
      ├───┤┌─┴─┐└╥┘┌─┐
q0_1: ┤ I ├┤ X ├─╫─┤M├
      └───┘└───┘ ║ └╥┘
c0: 2/═══════════╩══╩═
                 0  1

## Part 5
### Simulation

In [27]:
from qiskit_aer import AerSimulator

In [28]:
backend = AerSimulator()

In [29]:
from qiskit import transpile

In [30]:
transpiled_circuit = transpile(circuit, backend = backend)

In [31]:
from qiskit_ibm_runtime import SamplerV2 as Sampler

In [32]:
sampler = Sampler(mode = backend)
job = sampler.run([transpiled_circuit], shots = 1024)

In [33]:
job.status()

<JobStatus.DONE: 'job has successfully run'>

In [34]:
results = job.result()[0].data
outputs = {}
for key in results.__dict__:
 outputs[key] = results.__dict__[key].get_counts()

In [35]:
print(outputs)

{'c0': {'11': 504, '00': 520}}


In [36]:
circuit.draw()

┌───┐     ┌─┐   
q0_0: ┤ H ├──■──┤M├───
      ├───┤┌─┴─┐└╥┘┌─┐
q0_1: ┤ I ├┤ X ├─╫─┤M├
      └───┘└───┘ ║ └╥┘
c0: 2/═══════════╩══╩═
                 0  1

## Part 6

### Real backend

In [28]:
service.backends()

[<IBMBackend('ibm_brisbane')>,
 <IBMBackend('ibm_pittsburgh')>,
 <IBMBackend('ibm_fez')>,
 <IBMBackend('ibm_marrakesh')>,
 <IBMBackend('ibm_torino')>,
 <IBMBackend('ibm_kingston')>]

In [29]:
#backend = service.backend("ibm_kingston")

In [30]:
#transpiled_circuit = transpile(circuit, backend = backend)
#sampler = Sampler(mode = backend)
job = sampler.run([transpiled_circuit], shots = 1024)
job.status()

'QUEUED'

In [44]:
job.status()

'DONE'

In [45]:
results = job.result()[0].data
outputs = {}
for key in results.__dict__:
 outputs[key] = results.__dict__[key].get_counts()
print(outputs)

{'c0': {'011': 260, '001': 266, '010': 248, '000': 250}}


### Backend Used IBM KINGSTON


### Screenshot of Transpiled Circuit

In [46]:
transpiled_circuit.draw()

global phase: π
           ┌─────────┐┌────┐┌───────┐   ┌────┐┌─────────┐   ┌─┐
q0_1 -> 83 ┤ Rz(π/2) ├┤ √X ├┤ Rz(π) ├─■─┤ √X ├┤ Rz(π/2) ├───┤M├
           ├─────────┤├────┤├───────┤ │ ├────┤├─────────┤┌─┐└╥┘
q0_0 -> 96 ┤ Rz(π/2) ├┤ √X ├┤ Rz(π) ├─■─┤ √X ├┤ Rz(π/2) ├┤M├─╫─
           └─────────┘└────┘└───────┘   └────┘└─────────┘└╥┘ ║ 
     c0: 3/═══════════════════════════════════════════════╩══╩═
                                                          0  1

### Compare the Counts for Each State:

- **For state '001':**
  - Simulator: 239
  - Real Quantum Computer: 266
  - **Difference:** The real quantum computer produced 27 more counts for this state than the simulator.

- **For state '011':**
  - Simulator: 294
  - Real Quantum Computer: 260
  - **Difference:** The simulator produced 34 more counts for this state than the real quantum computer.

- **For state '000':**
  - Simulator: 249
  - Real Quantum Computer: 250
  - **Difference:** The counts are really close, but real quantum computer produces 1 more count.

- **For state '010':**
  - Simulator: 242
  - Real Quantum Computer: 248
  - **Difference:** The real quantum computer produced 6 more counts for this state.


## Bonus

In [54]:

# Create quantum and classical registers
bonus_qubits = QuantumRegister(6)
bonus_cbits = ClassicalRegister(4)

# Initialize the quantum circuit with the registers
bonus_circuit = QuantumCircuit(bonus_qubits, bonus_cbits)

# Apply Hadamard gates to qubits 0 to 5
bonus_circuit.h(bonus_qubits[0:6])

# Apply Z gate to qubit 5
bonus_circuit.z(bonus_qubits[5])

# Apply barriers to qubits 0 to 5
bonus_circuit.barrier(bonus_qubits[0:6])

# Apply CNOT gates
bonus_circuit.cx(bonus_qubits[0], bonus_qubits[5])
bonus_circuit.cx(bonus_qubits[2], bonus_qubits[5])
bonus_circuit.cx(bonus_qubits[3], bonus_qubits[5])

# Apply barriers again
bonus_circuit.barrier(bonus_qubits[0:6])

# Apply Hadamard gates again to qubits 0 to 4
bonus_circuit.h(bonus_qubits[0:5])

# Measure qubits 0 to 4 and store results in classical bits
bonus_circuit.measure(bonus_qubits[0:4], bonus_cbits[0:4])
bonus_circuit.measure(bonus_qubits[4], bonus_cbits[3])

# Draw the quantum circuit
bonus_circuit.draw()


┌───┐      ░                 ░ ┌───┐┌─┐            
q8_0: ┤ H ├──────░───■─────────────░─┤ H ├┤M├────────────
      ├───┤      ░   │             ░ ├───┤└╥┘┌─┐         
q8_1: ┤ H ├──────░───┼─────────────░─┤ H ├─╫─┤M├─────────
      ├───┤      ░   │             ░ ├───┤ ║ └╥┘┌─┐      
q8_2: ┤ H ├──────░───┼────■────────░─┤ H ├─╫──╫─┤M├──────
      ├───┤      ░   │    │        ░ ├───┤ ║  ║ └╥┘┌─┐   
q8_3: ┤ H ├──────░───┼────┼────■───░─┤ H ├─╫──╫──╫─┤M├───
      ├───┤      ░   │    │    │   ░ ├───┤ ║  ║  ║ └╥┘┌─┐
q8_4: ┤ H ├──────░───┼────┼────┼───░─┤ H ├─╫──╫──╫──╫─┤M├
      ├───┤┌───┐ ░ ┌─┴─┐┌─┴─┐┌─┴─┐ ░ └───┘ ║  ║  ║  ║ └╥┘
q8_5: ┤ H ├┤ Z ├─░─┤ X ├┤ X ├┤ X ├─░───────╫──╫──╫──╫──╫─
      └───┘└───┘ ░ └───┘└───┘└───┘ ░       ║  ║  ║  ║  ║ 
c8: 4/═════════════════════════════════════╩══╩══╩══╩══╩═
                                           0  1  2  3  3

In [57]:
bonus_backend = AerSimulator()
bonus_transpiled_circuit = transpile(bonus_circuit, backend = bonus_backend)
sampler = Sampler(mode = bonus_backend)
bonus_job = sampler.run([bonus_transpiled_circuit], shots = 1024)
bonus_job.status()


<JobStatus.RUNNING: 'job is actively running'>

In [58]:
bonus_job.status()

<JobStatus.DONE: 'job has successfully run'>

In [59]:
bonus_results = bonus_job.result()[0].data
outputs = {}
for key in bonus_results.__dict__:
 outputs[key] = bonus_results.__dict__[key].get_counts()

## Sharing My Results

In [60]:
print(outputs)

{'c8': {'0101': 1024}}
